sql/plan/indexed_in_subquery_filter.go

```diff
@@ -0,0 +1,156 @@
+)
+
+// NewIndexedInSubqueryFilter returns an IndexedInSubqueryFilter
+// sql.Node. The Node implements the semantics of `Filter(field IN
+// (SELECT ...), Child)`, but runs the subquery `SELECT` first, and
+// the makes repeated calls to `child.RowIter()` to get the matching
+// Child rows. Typically `child` should be an `IndexedTableAccess`,
+// and for this to be the right tradeoff, the results from `subquery`
+// should be much smaller than the unfiltered results from `child`.
+// `padding` is the number of `null` columns which will be appended to
+// an incoming `Row` in the `RowIter` call because calling `Eval` on
+// the `Subquery`, since the `Subquery` was originally expecting to
+// expect within the scope of something like `child`. `child` itself
+// should expect `RowIter()` calls with a single column `Row`, which
+// will be the results from the `Subquery`. `filterField` is a
+// `GetField` expression which will extract the field from `child`
+// results that should be matched against the `subquery` results; this
+// condition is still checked here because `child` is allowed to
+// return non-matching rows. `equals` true means this node will call
+// `subquery.Eval` and expect a single result, whereas `equals` false
+// means this node will call `subquery.EvalMultiple` and expect 0 or
+// more results.
+func NewIndexedInSubqueryFilter(subquery *Subquery, child sql.Node, padding int, filterField *expression.GetField, equals bool) sql.Node {
+	return &IndexedInSubqueryFilter{subquery, child, padding, filterField, equals}
+}
+
+type IndexedInSubqueryFilter struct {
+	Subquery *Subquery
+	Child    sql.Node
+	Padding  int
+	GetField *expression.GetField
+	Equals   bool
+}
+
+func (i *IndexedInSubqueryFilter) Resolved() bool {
+	return i.Subquery.Resolved() && i.Child.Resolved()
+}
+
+func (i *IndexedInSubqueryFilter) String() string {
+	pr := sql.NewTreePrinter()
+	_ = pr.WriteNode("IndexedInSubqueryFilter(%s IN (%s))", i.GetField, i.Subquery)
+	_ = pr.WriteChildren(i.Child.String())
+	return pr.String()
+}
+
+func (i *IndexedInSubqueryFilter) DebugString() string {
+	pr := sql.NewTreePrinter()
+	_ = pr.WriteNode("IndexedInSubqueryFilter(%s IN (%s))", sql.DebugString(i.GetField), sql.DebugString(i.Subquery))
+	_ = pr.WriteChildren(sql.DebugString(i.Child))
+	return pr.String()
+}
+
+func (i *IndexedInSubqueryFilter) Schema() sql.Schema {
+	return i.Child.Schema()
+}
+
+func (i *IndexedInSubqueryFilter) Children() []sql.Node {
+	return []sql.Node{i.Child}
+}
+
+func (i *IndexedInSubqueryFilter) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 1 {
+		return nil, sql.ErrInvalidChildrenNumber.New(i, len(children), 1)
+	}
+	return NewIndexedInSubqueryFilter(i.Subquery, children[0], i.Padding, i.GetField, i.Equals), nil
+}
+
+func (i *IndexedInSubqueryFilter) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	padded := make(sql.Row, len(row)+i.Padding)
```

Comment:
You sure this padding is necessary? A subquery expects to get a row for the outer scope. Why does it need nils on the end?

Reply:
If you have a getField expression that reaches into the nil part of the row given, it's not going to give you a correct result

Reply:
It's necessary because I didn't rewrite the subquery. It's going to append its rows onto the end of the row that comes in, and it's going to be allowed to do field accesses into the prefix and the suffix of that row. Based on the inspection, it shouldn't access these `nil` values though.

Reply:
OK, I get it now. The field indexes in the subquery are based off analysis of the subquery expression using the scope node, and you're not giving it this scope node. You're giving it a fake one you construct that has the same length. And these null vals won't get used because this node won't be constructed in that case.

Some time I'm going to have to clean up the analyzer to make scope management more principled / consistent.